### DISCUSSION TITLE
Product type not supported - What are my options?

### DISCUSSION THREAD

**Discussion by Grant3W** [amzn/selling-partner-api-models#3244] [General] [answered]
Hi all,

I'm currently implementing the SP API, but on both the Listings API and Feeds API, I am getting the following response for my listings.
```
"code": "4000003",
"severity": "ERROR",
"message": "The Amazon product type specified is invalid or not supported."
```

I am trying to use the `PLANT_SEED` product type, and I can see it's not listed when I search the `/definitions/2020-09-01/productTypes` endpoint. With all that in mind, what are my options for programmatically listing products using the SP-API for this product type?

Thanks.

**Comment by chapmanjw** [comment, marked as answer]
@Grant3W Until that product type is launched in the Definitions + Listings APIs, you can update common elements (like availability) with the Listings APIs with the `PRODUCT` product type. 

For full product creates prior to the product type launching in the Definitions + Listings APIs, you can use the `POST_PRODUCT_DATA` feed with the Feeds API: https://github.com/amzn/selling-partner-api-docs/blob/main/references/feeds-api/feedtype-values.md.

**Reply by KajdeMunter** [reply]
Using productType `PRODUCT` makes my feed stuck `IN_PROGRESS`. Do you have any idea how to fix this and/or why this happens? If I use the listings API to get the product_type I also get `4000003 The Amazon product type specified is invalid or not supported`.

My feed document:

```
{
  "header": {
    "sellerId": "...",
    "version": "2.0",
    "issueLocale": "nl_NL"
  },
  "messages": [
    {
      "messageId": 1,
      "sku": "5H-HRYL-HAZ7",
      "operationType": "PARTIAL_UPDATE",
      "productType": "PRODUCT",
      "attributes": {
        "fulfillment_availability": {
          "quantity": 10
        }
      }
    },
    {
      "messageId": 2,
      "sku": "TEST_SKU",
      "operationType": "PARTIAL_UPDATE",
      "productType": "PRODUCT",
      "attributes": {
        "fulfillment_availability": {
          "quantity": 20
        }
      }
    }
  ]
}
```

**Reply by chapmanjw** [reply]
You can use the Listings Items API `patchListingsItem` operation to update `fulfillment_availability` using the `PRODUCT` product type (the `putListingsItem` operation does not support creating new listings using the root `PRODUCT` product type, but you can use it for editing common attributes in the `patchListingsItem`). 

The JSON_LISTINGS_FEED is essentially turning your messages into these patch operations. However, the payload is incorrectly formatted and hitting a bug causing an internal error rather than a client error (we're working on fixing this part).

Whether you use the JSON_LISTINGS_FEED or the `patchListingsItem` operation, `fulfillment_availability` is a list, not a singular object. For example:

```
"fulfillment_availability":[ 
  {
    "quantity": 20,
    "fulfillment_channel_code":"DEFAULT"
  }
]
```

**Reply by KajdeMunter** [reply]
Awesome, thank you for the clarification!

I've now fixed it using the following payload:

```
{
  "header": {
    "sellerId": "...",
    "version": "2.0",
    "issueLocale": "nl_NL"
  },
  "messages": [
    {
      "messageId": 1,
      "sku": "5H-HRYL-HAZ7",
      "operationType": "PATCH",
      "productType": "PRODUCT",
      "patches": [
        {
          "op": "replace",
          "path": "/attributes/fulfillment_availability",
          "value": [
            {
              "fulfillment_channel_code": "DEFAULT",
              "quantity": 10
            }
          ]
        }
      ]
    }
  ]
}
```

**Reply by chapmanjw** [reply]
Side note: we fixed the bug that was causing the internal errors when the payload was malformed.

**Reply by mzbk** [reply]
> Awesome, thank you for the clarification!
> 
> I've now fixed it using the following payload:
> 
> ```
> {
>   "header": {
>     "sellerId": "...",
>     "version": "2.0",
>     "issueLocale": "nl_NL"
>   },
>   "messages": [
>     {
>       "messageId": 1,
>       "sku": "5H-HRYL-HAZ7",
>       "operationType": "PATCH",
>       "productType": "PRODUCT",
>       "patches": [
>         {
>           "op": "replace",
>           "path": "/attributes/fulfillment_availability",
>           "value": [
>             {
>               "fulfillment_channel_code": "DEFAULT",
>               "quantity": 10
>             }
>           ]
>         }
>       ]
>     }
>   ]
> }
> ```

What if we want to update multiple product quantities? Can we use the patchListingsItem method? **OR** for multiple we need to create the feed? @chapmanjw 
With patchListingsItem, I can update one product inventory I need to know how can i update for multiples and what is the endpoint for multiple.
Thank you.

**Reply by chapmanjw** [reply]
@mzbk The patchListingItem method accepts updates for one listing at a time. The JSON_LISTINGS_FEED accepts updates for multiple listings per submission. You create separate message entries for each SKU update.

**Reply by PhenomEY** [reply]
> @Grant3W Until that product type is launched in the Definitions + Listings APIs, you can update common elements (like availability) with the Listings APIs with the `PRODUCT` product type.
> 
> For full product creates prior to the product type launching in the Definitions + Listings APIs, you can use the `POST_PRODUCT_DATA` feed with the Feeds API: https://github.com/amzn/selling-partner-api-docs/blob/main/references/feeds-api/feedtype-values.md.

Is there any new information on when the productType will be finally available in the ListingsAPI?

**Comment by SatyamGupta1922Cedcommerce** [comment]
Whats is the productType for puListing Api?

**Reply by Philomath88** [reply]
@chapmanjw Can you advise on how to convert from the JSON (via getListingItem) to creating an appropriate XML feed? I have the individual logic for both but I have a hard time converting from JSON to XML (and to have the latter validate against the schemas). 

The purpose is to be able to relist items programmatically for not supported ListingItem types

Ideally, I would like to use an operating like getListingItem to get a XML from an existing product. This would make it soooo much easier to work with.
